File: app/src/main/java/com/upadhyay/hemant/NoteEx/activities/MainActivity.java
package com.upadhyay.hemant.NoteEx.activities;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.upadhyay.hemant.NoteEx.R;
import com.upadhyay.hemant.NoteEx.adapters.NotesAdapter;
import com.upadhyay.hemant.NoteEx.database.NotesDatabase;
import com.upadhyay.hemant.NoteEx.entities.Note;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    public static final int REQUEST_CODE_ADD_NOTE = 1;

    private RecyclerView notesRecyclerView;
    private List<Note> noteList;
    private NotesAdapter notesAdapter;

    private ImageView imageViewAddNoteFab;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();

        imageViewAddNoteFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivityForResult(
                        new Intent(getApplicationContext(), CreateNoteActivity.class),
                        REQUEST_CODE_ADD_NOTE
                );
            }
        });

        notesRecyclerView.setLayoutManager(
                new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        );
        noteList = new ArrayList<>();
        notesAdapter = new NotesAdapter(noteList);
        notesRecyclerView.setAdapter(notesAdapter);

        getNotes();
    }

    public void getNotes(){
        // Async task to get all the saved notes
        class GetNotesTask extends AsyncTask<Void, Void, List<Note>>{
            @Override
            protected List<Note> doInBackground(Void... voids) {

                return NotesDatabase
                        .getNotesDatabase(getApplicationContext())
                        .noteDao().getAllNotes();
            }

            @Override
            protected void onPostExecute(List<Note> notes) {
                super.onPostExecute(notes);

                Log.d("TAG", "note: " + notes.toString());
                Log.d("TAG", "noteList: " + noteList.toString());

                if(noteList.size() == 0){
                    noteList.addAll(notes);
                    notesAdapter.notifyDataSetChanged();
                }else{
                    noteList.add(0, notes.get(0));
                    notesAdapter.notifyItemInserted(0);
                }
                notesRecyclerView.smoothScrollToPosition(0);
            }


        }

        new GetNotesTask().execute();

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == REQUEST_CODE_ADD_NOTE && resultCode == RESULT_OK){
            getNotes();
        }

    }

    private void initViews() {
        imageViewAddNoteFab = findViewById(R.id.imageViewAddNoteFab);
        notesRecyclerView = findViewById(R.id.notesRecyclerView);
    }
}

File: app/src/main/java/com/upadhyay/hemant/NoteEx/activities/MainActivity.kt
package com.upadhyay.hemant.NoteEx.activities

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.upadhyay.hemant.NoteEx.adapters.NotesAdapter
import android.os.Bundle
import android.content.Intent
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import android.os.AsyncTask
import com.upadhyay.hemant.NoteEx.data.database.NotesDatabase
import android.util.Log
import android.widget.ImageView
import com.akshatbhuhagal.mynotes.R
import com.upadhyay.hemant.NoteEx.data.entities.NoteEntity
import java.util.ArrayList

class MainActivity : AppCompatActivity() {
    private var notesRecyclerView: RecyclerView? = null
    private var noteList: MutableList<NoteEntity>? = null
    private var notesAdapter: NotesAdapter? = null
    private var imageViewAddNoteFab: ImageView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initViews()
        imageViewAddNoteFab!!.setOnClickListener {
            startActivityForResult(
                Intent(applicationContext, CreateNoteActivity::class.java),
                REQUEST_CODE_ADD_NOTE
            )
        }
        notesRecyclerView!!.layoutManager =
            StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        noteList = ArrayList()
        notesAdapter = NotesAdapter(noteList)
        notesRecyclerView!!.adapter = notesAdapter
        notes
    }

    // Async task to get all the saved notes
    val notes: Unit
        get() {
            // Async task to get all the saved notes
            class GetNotesTask : AsyncTask<Void?, Void?, List<NoteEntity>>() {
                protected override fun doInBackground(vararg voids: Void): List<NoteEntity> {
                    return NotesDatabase
                        .getNotesDatabase(applicationContext)
                        .noteDao().allNotes
                }

                override fun onPostExecute(notes: List<NoteEntity>) {
                    super.onPostExecute(notes)
                    Log.d("TAG", "note: $notes")
                    Log.d("TAG", "noteList: " + noteList.toString())
                    if (noteList!!.size == 0) {
                        noteList!!.addAll(notes)
                        notesAdapter!!.notifyDataSetChanged()
                    } else {
                        noteList!!.add(0, notes[0])
                        notesAdapter!!.notifyItemInserted(0)
                    }
                    notesRecyclerView!!.smoothScrollToPosition(0)
                }
            }
            GetNotesTask().execute()
        }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_ADD_NOTE && resultCode == RESULT_OK) {
            notes
        }
    }

    private fun initViews() {
        imageViewAddNoteFab = findViewById(R.id.imageViewAddNoteFab)
        notesRecyclerView = findViewById(R.id.notesRecyclerView)
    }

    companion object {
        const val REQUEST_CODE_ADD_NOTE = 1
    }
}